crates/lib/src/qpu/api.rs

```diff
@@ -202,6 +205,59 @@ pub struct ExecutionOptions {
     #[doc = "The timeout to use for the request, defaults to 30 seconds. If set to `None`, then there is no timeout."]
     #[builder(default = "Some(Duration::from_secs(30))")]
     timeout: Option<Duration>,
+    #[doc = "If `true` and the user is authorized, jobs will bypass change protection against managed settings."]
+    #[builder(default = "None")]
+    api_options: Option<ApiExecutionOptions>,
+}
+
+/// Options avaialable when executing a job on a QPU, particular to the execution service's API.
+///
+/// Use [`Default`] to get a reasonable set of defaults, or start with [`ApiExecutionOptionsBuilder`]
+/// to build a custom set of options.
+#[derive(Builder, Clone, Debug, Default)]
+#[allow(clippy::module_name_repetitions)]
+pub struct ApiExecutionOptions {
+    /// the inner proto representation
+    inner: InnerApiExecutionOptions,
+}
+
+impl PartialEq for ApiExecutionOptions {
+    /// implement Eq
+    fn eq(&self, other: &Self) -> bool {
+        self.inner.bypass_settings_protection == other.inner.bypass_settings_protection
+    }
+}
+
+impl Eq for ApiExecutionOptions {}
+
+impl ApiExecutionOptions {
+    /// Get an [`ExecutionOptionsBuilder`] that can be used to build a custom [`ExecutionOptions`].
+    #[must_use]
+    pub fn builder() -> ApiExecutionOptionsBuilder {
+        ApiExecutionOptionsBuilder::default()
+    }
+
+    /// Get the configured `bypass_settings_protection` value.
+    #[must_use]
+    pub fn bypass_settings_protection(&self) -> bool {
+        self.inner.bypass_settings_protection
+    }
+}
+
+impl From<ApiExecutionOptions> for InnerApiExecutionOptions {
+    fn from(options: ApiExecutionOptions) -> Self {
+        options.inner
+    }
+}
+
+impl ApiExecutionOptionsBuilder {
+    /// Set the `bypass_settings_protection` value.
+    pub fn bypass_settings_protection(&mut self, bypass_settings_protection: bool) -> &mut Self {
+        self.inner
+            .get_or_insert(InnerApiExecutionOptions::default())
+            .bypass_settings_protection = bypass_settings_protection;
+        self
```

Comment:
From a user perspective, is the difference between `ExecutionOptions` and `APIExecutionOptions` clear? For example, would a user understand why "ConnectionStrategy" is an `ExecutionOption` but not an `APIExecutionOption`? I think I see why we've nested the builders, but I wonder if specifying each option in `ExecutionOptions` directly would be better? It looks like in either case we'll have to maintain a set of functions for each option, so maybe we should opt for an approach that offers a simpler and less verbose API?

For clarity, I'm proposing getting rid of ApiExeuctionOptions and instead:

```rs
pub struct ExecutionOptions {
    // .. current options
    bypass_settings_protection: bool
}

impl From<ExecutionOptions> for InnerExecutionApiOptions {
    fn from(options: ExecutionOptions) -> Self {
         InnerExecutionApiOptions { bypass_settings_protection: ExecutionOptions.bypass_settings_protection }
    }
}
```

Does that make sense? Maybe I'm missing other important advantages for nesting the options?

Reply:
I agree it is confusing, and I had them combined on the first go around as you suggested, but to allow these structs to evolve independently and to allow any `impl From<T> for ApiExecutionOptions`, this seems to be more flexible. 

I'm flexible on this though, so happy to make the change considering the ergonomics.

Reply:
I'm inclined to prefer a single type as well, but what's the connection to `impl From<T>`?

Reply:
Alternate implementation with a combined type: #416 - I do like that one a bit more

Reply:
Like with TranslationOptions, I'd like to allow users to use the auto-generated code directly, because there will certainly be a delay (and potentially errors) in adding newly added gRPC execution options to any manually maintained code here.

We need to allow users to:

* provide options which are used on the _client_ (like timeout) and those used by the service (like settings protection)
* provide the service options directly (via gRPC API client) or (optionally) via more ergonomic builder.

For the Python bindings, we've accepted for the time being that there's no direct access to the gRPC fields.

Reply:
Gotcha, I don't want to spin wheels on this, so happy to proceed with the current implementation, but I'll suggest one more alternative. My main gripe is with the nested builders, so what if we embed `ApiExecutionOptions` into `ExecutionOptions` directly and use custom setters to set options in the embedded API options, whilst still allowing it to be set directly if the user desires:

```rs
pub struct ExecutionOptions {
    // ... existing options
    
    #[builder(setter(custom))]
    bypass_settings_protection: bool
    #[builder(default)]
    api_options: ApiExecutionOptions
}

impl ExecutionOptionsBuilder {
    fn bypass_settings_protection(&mut self, value: bool) {
        self.api_options.bypass_settings_protection = value;
    }
}
```

Reply:
https://github.com/rigetti/qcs-sdk-rust/pull/418 makes the argument generic, which allows direct use of  the gRPC type, but makes usage a little more awkward overall

Reply:
I'm going to leave this as-is for now with the nested builders, I think it would be a more convenient interface but also blurs the line between which struct is which. We're not necessarily blocked from making the change in another MR though.